qadence/ml_tools/train_grad.py

```diff
@@ -3,25 +3,29 @@
 from typing import Callable
 
 from rich.progress import BarColumn, Progress, TaskProgressColumn, TextColumn, TimeRemainingColumn
-from torch import Tensor
+from torch import cuda, set_default_device
+from torch import device as torchdevice
 from torch.nn import Module
 from torch.optim import Optimizer
 from torch.utils.data import DataLoader
 from torch.utils.tensorboard import SummaryWriter
 
 from qadence.logger import get_logger
 from qadence.ml_tools.config import TrainConfig
-from qadence.ml_tools.data import DictDataLoader
+from qadence.ml_tools.data import DataLoaderType, DictDataLoader, dataloader_to_device
 from qadence.ml_tools.optimize_step import optimize_step
 from qadence.ml_tools.printing import print_metrics, write_tensorboard
 from qadence.ml_tools.saveload import load_checkpoint, write_checkpoint
 
 logger = get_logger(__name__)
 
+DEVICE = torchdevice("cuda") if cuda.is_available() else torchdevice("cpu")
+set_default_device(DEVICE)
+
 
 def train(
     model: Module,
-    dataloader: DictDataLoader | DataLoader | list[Tensor] | tuple[Tensor, Tensor] | None,
+    dataloader: DataLoaderType,
     optimizer: Optimizer,
     config: TrainConfig,
     loss_fn: Callable,
```

Comment:
In the `ml_tools` it is fine to actually keep the calls to `.to()`, I think. This is a high-routine that has complete control on everything (data, models, optimizer, etc. etc.) so we can be sure that everything is indeed sent to the right device.

Reply:
i did that now